experiments/compositions/composition/internal/controller/expander_reconciler.go

```diff
@@ -329,65 +332,24 @@ func (r *ExpanderReconciler) SetupWithManager(mgr ctrl.Manager, cr *unstructured
 		Complete(r)
 }
 
-func (r *ExpanderReconciler) getExpanderVersion(ctx context.Context, inputExpanderVersion string, expanderType string) (string, string, error) {
+func (r *ExpanderReconciler) getExpanderVersion(ctx context.Context, inputExpanderVersion string, expanderType string) (string, *compositionv1alpha1.ExpanderVersion, error) {
 	logger := log.FromContext(ctx)
 
-	var expanderVersionCR compositionv1alpha1.ExpanderVersion
-	var expanderVersion, expanderRegistry string
-	var availableVersions []string
-	err := r.Client.Get(ctx, types.NamespacedName{Name: "composition-" + expanderType, Namespace: "composition-system"}, &expanderVersionCR)
+	var ev compositionv1alpha1.ExpanderVersion
+	err := r.Client.Get(ctx, types.NamespacedName{Name: "composition-" + expanderType, Namespace: "composition-system"}, &ev)
 	if err != nil {
 		// The CR should be created before the specified expander can be used.
-		return "", "", err
-	}
-	expanderRegistry = expanderVersionCR.Spec.ImageRegistry
-	availableVersions = expanderVersionCR.Spec.ValidVersions
-	SemVerVersions := make([]*semver.Version, len(availableVersions))
-	for i, r := range availableVersions {
-		v, err := semver.NewVersion(r)
-		if err != nil {
-			logger.Info("Error parsing version: %s", err)
-		}
-		SemVerVersions[i] = v
-	}
-	sort.Sort(semver.Collection(SemVerVersions))
-	logger.Info("inputexpanderverions", "current", inputExpanderVersion)
-	// Currently the version will default to latest if not set.
-	if inputExpanderVersion != "latest" {
-		// Verify if the input version is valid.
-		semVersion, err := semver.NewVersion(inputExpanderVersion)
-		if err != nil {
-			return "", "", fmt.Errorf("invalidExpanderVersion")
-		}
-		if !isValidVersion(SemVerVersions, semVersion) {
-			return "", "", fmt.Errorf("invalidExpanderVersion")
-		}
-		expanderVersion = inputExpanderVersion
-	} else {
-		// The existing semver package sort only supports pure numbers.
-		expanderVersion = "v" + SemVerVersions[len(availableVersions)-1].String() + ".alpha"
+		return "", nil, err
 	}
-	return expanderVersion, expanderRegistry, nil
-}
 
-func isValidVersion(availableVersions []*semver.Version, version *semver.Version) bool {
-	var low, high, mid int
-	low = 0
-	high = len(availableVersions) - 1
-	for low <= high {
-		mid = low + (high-low)/2
+	if ev.Status.VersionMap == nil {
```

Comment:
How does this version map gets populated? And what are the values that it is mapping?

Reply:
this is generated from the validversions. this is the status field that is populated by the expanderversion reconciler.
https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/experiments/compositions/composition/internal/controller/expanderversion_controller.go#L111

examples:

```
❯ make deploy-kind
❯ kubectl get expanderversion -A -o yaml
apiVersion: v1
items:
- apiVersion: composition.google.com/v1alpha1
  kind: ExpanderVersion
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"composition.google.com/v1alpha1","kind":"ExpanderVersion","metadata":{"annotations":{},"name":"composition-gjinja2","namespace":"composition-system"},"spec":{"type":"grpc","validVersions":["v0.0.1","v0.0.0"]}}
    creationTimestamp: "2024-05-16T22:38:15Z"
    generation: 1
    name: composition-gjinja2
    namespace: composition-system
    resourceVersion: "693"
    uid: febc6da0-bdda-4c24-9d63-a09e215904b0
  spec:
    type: grpc
    validVersions:
    - v0.0.1
    - v0.0.0
  status:
    versionMap:
      latest: composition-gjinja2-v0-0-1
      v0.0.0: composition-gjinja2-v0-0-0
      v0.0.1: composition-gjinja2-v0-0-1
- apiVersion: composition.google.com/v1alpha1
  kind: ExpanderVersion
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"composition.google.com/v1alpha1","kind":"ExpanderVersion","metadata":{"annotations":{},"name":"composition-jinja2","namespace":"composition-system"},"spec":{"imageRegistry":"gcr.io/krmapihosting-release","type":"job","validVersions":["v0.0.1","v0.0.0"]}}
    creationTimestamp: "2024-05-16T22:38:15Z"
    generation: 1
    name: composition-jinja2
    namespace: composition-system
    resourceVersion: "694"
    uid: 5a8fe82d-0798-4133-a5ed-f7926ae2ff34
  spec:
    imageRegistry: gcr.io/krmapihosting-release
    type: job
    validVersions:
    - v0.0.1
    - v0.0.0
  status:
    versionMap:
      latest: gcr.io/krmapihosting-release/expander-jinja2:v0.0.1
      v0.0.0: gcr.io/krmapihosting-release/expander-jinja2:v0.0.0
      v0.0.1: gcr.io/krmapihosting-release/expander-jinja2:v0.0.1
kind: List
metadata:
  resourceVersion: ""
```